pkg/config/reloader.go

```diff
@@ -0,0 +1,166 @@
+	filename string,
+	reloaders []ComponentReloader,
+) (*ConfigReloader, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to establish config watcher", "err", err)
+		return nil, err
+	}
+
+	if err := watcher.Add(filename); err != nil {
+		level.Error(logger).Log("msg", "failed to start watching config file", "err", err, "path", filename)
+		return nil, err
+	}
+
+	r := &ConfigReloader{
+		logger:   logger,
+		filename: filename,
+		watcher:  watcher,
+
+		reloaders: reloaders,
+
+		triggerReload: make(chan struct{}, 1),
+
+		configSuccess: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_successful",
+			Help: "Whether the last configuration reload attempt was successful.",
+		}),
+		configSuccessTime: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_success_timestamp_seconds",
+			Help: "Timestamp of the last successful configuration reload.",
+		}),
+	}
+
+	reg.MustRegister(
+		r.configSuccess,
+		r.configSuccessTime,
+	)
+
+	return r, nil
+}
+
+func (r *ConfigReloader) watchFile() {
+	for {
+		select {
+		case event, ok := <-r.watcher.Events:
+			if !ok {
+				level.Debug(r.logger).Log("msg", "config file watcher events channel closed. exiting goroutine.")
+				return
+			}
+			if event.Op&fsnotify.Write == fsnotify.Write {
+				level.Debug(r.logger).Log("msg", "config file has been modified")
+				r.triggerReload <- struct{}{}
+			}
+		case err, ok := <-r.watcher.Errors:
+			if !ok {
+				level.Debug(r.logger).Log("msg", "config file watcher errors channel closed. exiting goroutine.")
+				return
+			}
+			level.Error(r.logger).Log("msg", "error encountered while watching config file", "err", err)
+		}
+	}
+}
+
+func (r *ConfigReloader) reloadFile() (err error) {
+	start := time.Now()
+	timings := []interface{}{}
+	level.Info(r.logger).Log("msg", "loading configuration file", "filename", r.filename)
+
+	defer func() {
+		if err == nil {
+			r.configSuccess.Set(1)
+			r.configSuccessTime.SetToCurrentTime()
+		} else {
+			r.configSuccess.Set(0)
+		}
+	}()
+
+	cfg, err := LoadFile(r.filename)
+	if err != nil {
+		return fmt.Errorf("failed to load configuration: %w", err)
+	}
+
+	if err = cfg.Validate(); err != nil {
+		return fmt.Errorf("parsed configuration invalid (--config-path=%q): %w", r.filename, err)
+	}
+
+	failed := false
+	for _, rl := range r.reloaders {
+		rstart := time.Now()
+		if err := rl.Reloader(cfg); err != nil {
+			level.Error(r.logger).Log("msg", "failed to apply configuration", "err", err)
+			failed = true
+		}
+		timings = append(timings, rl.Name, time.Since(rstart))
+	}
+	if failed {
+		return fmt.Errorf("one or more errors occurred while applying the new configuration (--config-path=%q)", r.filename)
+	}
+
+	l := []interface{}{"msg", "completed loading of configuration file", "filename", r.filename, "totalDuration", time.Since(start)}
+	level.Info(r.logger).Log(append(l, timings...)...)
```

Comment:
```suggestion
	level.Debug(r.logger).Log(append(l, timings...)...)
```
nit: Debug would be a better fit, I think.

Reply:
IMHO Info is the right level, it is the only log informing the user their config has been reloaded (by default). When re-configuring Prometheus/Alertmanager, I often look for this message

Reply:
I don't have a strong opinion on this. I think If everything works well, logging is really necessary. We already use logs extensively (thanks to me :grin:), it just creates noise.

Reply:
My opinion is strong. As a user, I need this feedback when updating the config file to know Parca reacted to it, it can also be used to identify issues that happen after. Enabling debug level would drown it in an insane amount of logs. From my experience with Prometheus/Alertmanager, it will only log every few days in most cases

Reply:
Alright. Sounds good to me.